baseplate/lib/timer.py

```diff
@@ -0,0 +1,40 @@
+import time
+
+from datetime import timedelta
+from threading import Thread
+from typing import Callable
+from typing import Optional
+
+
+class Timer:
+    """A restartable timer that execute an action after a timeout or at an interval."""
+
+    def __init__(
+        self, action: Callable[[], None], interval: timedelta, is_repeating: bool = False,
+    ) -> None:
+        self.action = action
+        self.interval = interval
+        self.is_repeating = is_repeating
+        self._thread: Optional[Thread] = None
+
+    def start(self) -> None:
+        if self.is_running():
+            return
+        self._thread = Thread(target=self._run)
+
+    def stop(self) -> None:
+        self._thread = None
+
+    def _run(self) -> None:
+        thread = self._thread
+        while self.is_running() and self._thread == thread:
+            time.sleep(self.interval.total_seconds())
+            self.action()
```

Comment:
I think we need to check `if self.is_running() and self._thread == thread:` here as well in case it was `stopped` while it was sleeping.  I _thiiiiink_ we also might want to use `is` rather than `==` since they should be the exact same object and that's what we really care about?

```python
    def _run(self) -> None:
        thread = self._thread
        while self.is_running() and self._thread is thread:
            time.sleep(self.interval.total_seconds())
            if !self.is_running() or self._thread is not thread:
                return
            self.action()
            if not self.is_repeating:
                self.stop()
```

If you want to pull that check into a function, maybe something like:

```python
def _is_active_thread(self, thread: threading.Thread) -> bool:
    return self.is_running() and self._thread is thread
```